aptos/contracts/sources/pyth.move

```diff
@@ -0,0 +1,727 @@
+                let price_identifier = price_feed::get_price_identifier(price_feed);
+                state::set_latest_price_info(
+                    *price_identifier,
+                    update,
+                )
+            }
+        };
+        vector::destroy_empty(updates);
+    }
+
+    /// Update the cached price feeds with the data in the given vaa_bytes payload, using
+    /// update_price_feeds(). However, this function will only have an effect if any of the
+    /// prices in the update are fresh. The price_identifiers and publish_times paramaters
+    /// are used to determine if the update is fresh without doing any serialisation or verification
+    /// of the VAA, potentially saving time and gas. If the update contains no fresh data, this function
+    /// will revert with error::no_fresh_data(). 
+    /// 
+    /// For a given price update i in the batch, that price is considered fresh if the current cached 
+    /// price for price_identifiers[i] is older than publish_times[i].
+    public entry fun update_price_feeds_if_fresh(
+        vaa_bytes: vector<u8>,
+        price_identifiers: vector<PriceIdentifier>,
+        publish_times: vector<u64>,
+        fee: Coin<AptosCoin>) {
+
+        assert!(vector::length(&price_identifiers) == vector::length(&publish_times),
+            error::invalid_publish_times_length());
+
+        let fresh_data = false;
+        let i = 0;
+        while (i < vector::length(&publish_times)) {
+            let price_identifier = vector::borrow(&price_identifiers, i);
+            if (!state::price_info_cached(*price_identifier)) {
+                fresh_data = true;
+                break
+            };
+
+            let price_feed = price_info::get_price_feed(&state::get_latest_price_info(*price_identifier));
+            let cached_timestamp = price::get_timestamp(&price_feed::get_price(price_feed));
+            if (cached_timestamp < *vector::borrow(&publish_times, i)) {
+                fresh_data = true;
+                break
+            };
+
+            i = i + 1;
+        };
+
+        assert!(fresh_data, error::no_fresh_data());
+        update_price_feeds(vaa_bytes, fee);
+    }
+
+    /// Determine if the given price update is "fresh": we have nothing newer already cached for that
+    /// price feed.
+    fun is_fresh_update(update: &PriceInfo): bool {
+        // Get the timestamp of the update's current price
+        let price_feed = price_info::get_price_feed(update);
+        let update_timestamp = price::get_timestamp(&price_feed::get_price(price_feed));
+
+        // Get the timestamp of the cached data for the price identifier
+        let price_identifier = price_feed::get_price_identifier(price_feed);
+        if (!price_feed_exists(*price_identifier)) {
+            return true
+        };
+        let cached_timestamp = price::get_timestamp(
+            &price_feed::get_price(price_info::get_price_feed(&state::get_latest_price_info(*price_identifier))));
```

Comment:
@tompntn shouldn't you use unsafe versions here? Even for the update timestamp (maybe it's old but still newer).

Reply:
This uses `get_latest_price_info`, which is an internal getter function that has no timestamp check. However, using get_price_unsafe would make this code simpler 😄

Reply:
i don't know. i see get_price both as contract interface and method on struct. so probably i'm wrong here.